design-docs/cameraserver.adoc

```diff
@@ -0,0 +1,419 @@
+listening to a particular input, the library may pause or disconnect
+from a input to save processor and I/O resources.  The library
+autonomously handles camera disconnects/reconnects by simply pausing
+and resuming the image stream.
+
+* Outputs listen to a particular input's event, grab the latest
+image, and forward it to its destination in the appropriate format.
+Similarly to inputs, if a particular output is inactive (e.g. no
+client is connected to a configured M-JPEG over HTTP server), the
+library will disable parts of its processing to save processor
+resources.
+
+* In addition, user code (such as that used in a FRC robot program)
+can act as either a input (providing processed frames) or as an
+output (receiving a frame from a camera).
+
+The library distinguishes "frames" from "images" as some types of
+cameras (such as depth cameras or binocular vision cameras) can
+provide multiple images per frame (the library refers to these as
+numbered "channels").  Output modules are only capable of sending
+single images, so it's necessary to configure an output module to
+select which channel of the input should be used for the output
+stream.
+
+The library is designed to run as separate threads within the user
+process.  This is primarily done for efficiency reasons in the case of
+user image processing, but also ensures a clean, consistent runtime
+environment each time the user program is run.
+
+[[implementation]]
+== Implementation
+
+For performance reasons, the library is implemented in {cpp} and
+provides both a C and {cpp} API.  A Java API is provided via JNI.
+
+=== Dependencies
+
+* OpenCV
+* WPILib Utility Library (TBR: or NTcore)
+
+=== Portability
+
+While the primary target platform is Linux (both embedded and
+desktop), the library also provides Windows support to enable use
+cases such as off-robot simulation or image processing testing.
+
+[[api]]
+== API
+
+The C API prefixes all names with `CS_`.
+
+The C API uses an opaque handle based approach.  While handles do not
+have unique types (all handles are `typedef int` in the API), they are
+globally unique in the library, and the library can detect bugs such
+as interchanging of a input and an output handle.
+
+Handle types:
+
+* CS_Input
+
+* CS_Output
+
+* CS_InputListener
+
+* CS_OutputListener
+
+The API for cameras is based on that provided by OpenCV.  This is
+intentional, as on Linux systems the functions will largely be
+passthrough (at least for USB cameras).
+
+=== Input Management
+
+==== CreateUSBCameraDev
+
+[source,c]
+CS_Input CS_CreateUSBCameraDev(int dev);
+
+Connects to a USB camera based on device number (e.g. on Linux,
+`/dev/videoN`).
+
+==== CreateUSBCameraPath
+
+[source,c]
+CS_Input CS_CreateUSBCameraPath(const char* path);
+
+Connects to a USB camera at the specified path.  Using this can
+resolve ambiguity when using multiple USB cameras.
+
+On Linux, this can either be `/dev/videoN` (note: this does not solve
```

Comment:
Note that OpenCV doesn't support symlinks without https://github.com/Itseez/opencv/pull/6391 (should be in 3.2+)